cloudinit/cmd/main.py

```diff
@@ -334,6 +335,68 @@ def _should_bring_up_interfaces(init, args):
     return not args.local
 
 
+def _should_wait_via_cloud_config(
+    raw_config: Optional[Union[str, bytes]]
+) -> bool:
+    if not raw_config:
+        return False
+
+    # If our header is anything other than #cloud-config, wait
+    possible_header: Union[bytes, str] = raw_config.strip()[:13]
+    if isinstance(possible_header, str):
+        decoded_header = possible_header
+    elif isinstance(possible_header, bytes):
+        try:
+            decoded_header = possible_header.decode("utf-8")
+        except UnicodeDecodeError:
+            return True
+    if not decoded_header.startswith("#cloud-config"):
+        return True
+
+    try:
+        userdata_yaml = yaml.safe_load(raw_config)
+    except Exception as e:
+        log_with_downgradable_level(
+            logger=LOG,
+            version="24.4",
+            requested_level=logging.WARNING,
+            msg="Unexpected failure parsing userdata: %s",
+            args=e,
+        )
+        return True
+
+    # These all have the potential to require network access, so we should wait
+    if "write_files" in userdata_yaml:
+        return any("source" in item for item in userdata_yaml["write_files"])
+    return bool(
+        userdata_yaml.get("bootcmd")
+        or userdata_yaml.get("random_seed", {}).get("command")
+        or userdata_yaml.get("mounts")
+    )
+
+
+def _should_wait_on_network(datasource: Optional[sources.DataSource]) -> bool:
+    """Determine if we should wait on network connectivity for cloud-init.
+
+    We need to wait if:
+    - We have no datasource
+    - We have user data that is anything other than cloud-config
+      - This can likely be further optimized in the future to include
+        other user data types
+    - We have user data that requires network access
+    """
+    if not datasource:
+        return True
+    return any(
+        _should_wait_via_cloud_config(config)
+        for config in [
+            datasource.get_userdata_raw(),
+            datasource.get_vendordata_raw(),
+            datasource.get_vendordata2_raw(),
+        ]
```

Comment:
Can we make this a generator expression without having to calculate the truthy return values from userdata vedordata and vendordarta2? Then we can early exit on the first case of userdata/vendordata containing write_files or un-decodable file content without having to perform the extra calculations.